### PR TITLE
feat: add related connections and nested view to datasource table.

### PR DESCRIPTION
**Pull request by weiwang118**
### Description
This pr implement new design for data source table in Admin->Data Sources Page:
- Add two fields related connections and type to the table.
- Add a Type Filter besides search bar
- Update direct query connections table with nested view showing direct queries connections sublist.
- Show Delete Connections button when selecting data sources.

## Screenshot
Screenshot 2024-08-29 at 00 03 05
Screenshot 2024-08-29 at 00 03 21
Screenshot 2024-08-29 at 00 03 40
Screenshot 2024-08-29 at 00 03 54

## Changelog

- skip

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by kamingleung** [comment]
@weiwang118 
On the Direct query tab, if a OpenSearch data source contains 0 related/nested connections, we should not show them here. Thanks!

cc @kgcreative

**Comment by weiwang118** [comment]
> @weiwang118 On the Direct query tab, if a OpenSearch data source contains 0 related/nested connections, we should not show those OpenSearch data sources in this tab. Thanks! image
> 
> cc @kgcreative

Hi, thanks for review. Already updated the commit with new change that only show data sources with 1 or more related connections in direct query connections tab.
Screenshot 2024-08-30 at 00 44 29

**Comment by kamingleung** [comment]
Hi @weiwang118 Thanks for a quick update: 
Can the nested data source names be indented like into the mockups? It would be difficult for users to differentiate nested items without indention. 
image

File: src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx [hunk 1]
```diff
@@ -80,8 +81,11 @@ export const DataSourceTable = ({ history }: RouteComponentProps) => {
   const fetchDataSources = () => {
     setIsLoading(true);
     getDataSources(savedObjects.client)
-      .then((response: DataSourceTableItem[]) => {
-        setDataSources(response);
+      .then(async (response: DataSourceTableItem[]) => {
+        return await fetchDataSourceConnections(response, http, notifications);
+      })
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx:86]: Any reason for not simply chaining the promise?
```suggestion
      .then((response: DataSourceTableItem[]) =>
        fetchDataSourceConnections(response, http, notifications);
      )
```

File: src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx [hunk 2]
```diff
@@ -96,46 +100,55 @@ export const DataSourceTable = ({ history }: RouteComponentProps) => {
...
-
-  const renderToolsRight = () => {
-    return canManageDataSource ? (
-      <EuiFlexItem key="delete" grow={false}>
-        {renderDeleteButton()}
-      </EuiFlexItem>
-    ) : null;
-  };
+  const renderToolsLeft = useCallback(() => {
+    return selectedDataSources.length > 0 && !confirmDeleteVisible
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx:104]: Why would we hide the button when the confirm is visible?

File: src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx [hunk 3]
```diff
@@ -96,46 +100,55 @@ export const DataSourceTable = ({ history }: RouteComponentProps) => {
...
       incremental: true,
       schema: {
         fields: { title: { type: 'string' } },
       },
     },
+    filters: [
+      {
+        type: 'field_value_selection',
+        field: 'type',
+        name: 'Type',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx:134]: Please use i18n.

File: src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx [hunk 4]
```diff
@@ -96,46 +100,55 @@ export const DataSourceTable = ({ history }: RouteComponentProps) => {
...
+      },
+    ],
   };
 
   /* Table columns */
   const columns = [
     {
       field: 'title',
-      name: 'Title',
+      name: 'Data source',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx:151]: Please use i18n.

File: src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx [hunk 5]
```diff
@@ -160,6 +173,11 @@ export const DataSourceTable = ({ history }: RouteComponentProps) => {
       dataType: 'string' as const,
       sortable: ({ title }: { title: string }) => title,
     },
+    {
+      field: 'type',
+      name: 'Type',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx:178]: Please use i18n.

File: src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx [hunk 6]
```diff
@@ -170,6 +188,13 @@ export const DataSourceTable = ({ history }: RouteComponentProps) => {
       dataType: 'string' as const,
       sortable: ({ description }: { description: string }) => description,
     },
+    {
+      field: 'relatedConnections',
+      name: 'Related connections',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx:193]: Please use i18n.

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 7]
```diff
@@ -3,145 +3,215 @@
...
+        // if yes, then set the first existing datasource as default datasource.
+        setDefaultDataSource();
       })
-      .catch((err) => {
-        notifications.toasts.addDanger('Could not fetch data sources');
+      .catch(() => {
+        notifications.toasts.addDanger(
+          i18n.translate('dataSourcesManagement.directQueryTable.deleteDataSourceFailMsg', {
+            defaultMessage:
+              'Error occurred while deleting selected records for Data sources. Please try it again',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:107]: This message seems off. How about:
```suggestion
              'An error occurred while attempting to delete the selected data sources. Please try again.',
```

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 8]
```diff
@@ -3,145 +3,215 @@
...
+      for (const dataSource of selectedDataSources) {
+        if (getDefaultDataSourceId(uiSettings) === dataSource.id) {
+          await setFirstDataSourceAsDefault(savedObjects.client, uiSettings, true);
+        }
+      }
+    } catch (e) {
+      notifications.toasts.addDanger(
+        i18n.translate('dataSourcesManagement.directQueryTable.setDefaultDataSourceFailMsg', {
+          defaultMessage:
+            'Unable to find a default datasource. Please set a new default datasource.',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:81]: This message sounds a bit off. How about:
```suggestion
            'No default data source has been set. Please select a new default data source.',
```

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 9]
```diff
@@ -3,145 +3,215 @@
...
+      </EuiConfirmModal>
+    ) : null;
+  };
 
-      const endpoint =
-        featureFlagStatus && selectedDataSourceId !== undefined
-          ? `${DATACONNECTIONS_BASE}/${connectionName}/dataSourceMDSId=${selectedDataSourceId}`
-          : `${DATACONNECTIONS_BASE}/${connectionName}`;
+  const renderToolsLeft = useCallback(() => {
+    return selectedDataSources.length > 0 && !isModalVisible
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:162]: Hiding the tools when modal is visible doesn't feel right.

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 10]
```diff
@@ -3,145 +3,215 @@
...
+      ? [
+          <EuiSmallButton
+            color="danger"
+            onClick={() => setIsModalVisible(true)}
+            data-test-subj="deleteDataSourceConnections"
+          >
+            Delete {selectedDataSources.length || ''}{' '}
+            {selectedDataSources.length ? 'connection' : ''}
+            {selectedDataSources.length >= 2 ? 's' : ''}
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:171]: We have `selectedDataSources.length > 0` in the path.
```suggestion
            <FormattedMessage
              id="dataSourcesManagement.directQueryTable.deleteToolLabel"
              defaultMessage="{selectionSize, plural, one {Delete # connection} other {Delete # connections}}"
              values={{ selectionSize: selectedDataSources.length }}
            />
```

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 11]
```diff
@@ -3,145 +3,215 @@
...
-        });
-    },
-    [http, notifications.toasts, selectedDataSourceId, featureFlagStatus, data]
-  );
+  const fetchDataSources = useCallback(() => {
+    setIsLoading(true);
+
+    getDataSources(savedObjects.client)
+      .then(async (response: DataSourceTableItem[]) => {
+        return await fetchDataSourceConnections(response, http, notifications);
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:202]: We don't need async and await here

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 12]
```diff
@@ -3,145 +3,215 @@
...
+      })
+      .then((finalData) => {
+        setData(finalData.filter((item) => item.relatedConnections?.length > 0));
+      })
+      .catch(() => {
+        setData([]);
+        notifications.toasts.addDanger('Could not fetch data sources');
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:209]: Please use i18n

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 13]
```diff
@@ -151,190 +221,88 @@ export const ManageDirectQueryDataConnectionsTable: React.FC<ManageDirectQueryDa
...
-        setIntegrationsFlyout(
-          <InstallIntegrationFlyout
-            closeFlyout={() => setShowIntegrationsFlyout(false)}
-            datasourceType={datasource.connectionType}
-            datasourceName={datasource.name}
-            http={http}
-          />
+      width: '25%',
+      field: 'title',
+      name: 'Data source',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:241]: Please use i18n

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 14]
```diff
@@ -151,190 +221,88 @@ export const ManageDirectQueryDataConnectionsTable: React.FC<ManageDirectQueryDa
...
-  };
-
-  const tableColumns = [
-    {
-      field: 'name',
-      name: 'Name',
-      sortable: true,
+      width: '15%',
+      field: 'type',
+      name: 'Type',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:259]: Please use i18n

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 15]
```diff
@@ -151,190 +221,88 @@ export const ManageDirectQueryDataConnectionsTable: React.FC<ManageDirectQueryDa
...
-        </EuiFlexGroup>
-      ),
     },
     {
-      field: 'status',
-      name: 'Status',
-      sortable: true,
+      width: '35%',
+      field: 'description',
+      name: 'Description',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:265]: Please use i18n

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 16]
```diff
@@ -151,190 +221,88 @@ export const ManageDirectQueryDataConnectionsTable: React.FC<ManageDirectQueryDa
...
+      mobileOptions: {
+        show: false,
+      },
     },
     {
-      field: 'actions',
-      name: 'Actions',
-      actions,
+      field: 'relatedConnections',
+      name: 'Related connections',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:273]: Please use i18n

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 17]
```diff
@@ -151,190 +221,88 @@ export const ManageDirectQueryDataConnectionsTable: React.FC<ManageDirectQueryDa
...
+    compressed: true,
+    filters: [
+      {
+        type: 'field_value_selection',
+        field: 'type',
+        name: 'Type',
+        multiSelect: 'or',
+        options: Array.from(new Set(data.map(({ type }) => type).filter(Boolean))).map((type) => ({
+          value: type!,
+          name: type!,
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:294]: is the `name` displayed on the UI? If yes, we should find a way to use i18n for it.

File: src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx [hunk 18]
```diff
@@ -151,190 +221,88 @@ export const ManageDirectQueryDataConnectionsTable: React.FC<ManageDirectQueryDa
...
+    toolsLeft: renderToolsLeft(),
+    box: {
+      incremental: true,
+    },
+    compressed: true,
+    filters: [
+      {
+        type: 'field_value_selection',
+        field: 'type',
+        name: 'Type',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/direct_query_data_sources_components/direct_query_data_connection/manage_direct_query_data_connections_table.tsx:290]: Please use i18n

File: src/plugins/data_source_management/public/components/utils.ts [hunk 19]
```diff
@@ -39,12 +42,77 @@ import {
...
+import { DATACONNECTIONS_BASE } from '../constants';
+
+export const getDirectQueryConnections = async (dataSourceId: string, http: HttpSetup) => {
+  const endpoint = `${DATACONNECTIONS_BASE}/dataSourceMDSId=${dataSourceId}`;
+  const res = await http.get(endpoint);
+  const directQueryConnections: DataSourceTableItem[] = res.map(
+    (dataConnection: DirectQueryDatasourceDetails) => ({
+      id: `${dataSourceId}-${dataConnection.name}`,
+      title: dataConnection.name,
+      type: dataConnection.connector === 'S3GLUE' ? 'Amazon S3' : 'Prometheus',
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/utils.ts:54]: This is not scalable. It would be better to check the `connector`, filter out incompatible ones, and then show the correct type.

Think of a scenario where a future version of OSD is used to configure connectors and then OSD 2.17 is used by some user. OSD 2.17 will assume those connectors are Prometheus connectors.

File: src/plugins/data_source_management/public/components/utils.ts [hunk 20]
```diff
@@ -39,12 +42,77 @@ import {
...
+) => {
+  try {
+    const directQueryConnectionsPromises = dataSources.map((ds) =>
+      getDirectQueryConnections(ds.id, http!).catch(() => [])
+    );
+    const directQueryConnectionsResult = await Promise.all(directQueryConnectionsPromises);
+    const directQueryConnections = directQueryConnectionsResult.flat();
+    return mergeDataSourcesWithConnections(dataSources, directQueryConnections);
+  } catch (error) {
+    notifications?.toasts.addDanger('Cannot fetch direct query connections');
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/utils.ts:99]: i18n please

File: src/plugins/data_source_management/public/types.ts [hunk 21]
```diff
@@ -39,11 +39,20 @@ export interface DataSourceManagementContext {
...
 }
 
+export enum DataSourceConnectionType {
+  OpenSearchConnection,
+  DirectQueryConnection,
+}
+
 export interface DataSourceTableItem {
   id: string;
+  type: string | undefined;
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/types.ts:49]: Any reason for not doing
```suggestion
  type?: string;
```

File: yarn.lock
Comment (AMoo-Miki) [inline yarn.lock:1]: Please undo the changes to the lockfile.
